models/record.go

```diff
@@ -0,0 +1,241 @@
+	TlsaSelector     uint8             `json:"tlsaselector,omitempty"`
+	TlsaMatchingType uint8             `json:"tlsamatchingtype,omitempty"`
+	TxtStrings       []string          `json:"txtstrings,omitempty"` // TxtStrings stores all strings (including the first). Target stores only the first one.
+	R53Alias         map[string]string `json:"r53_alias,omitempty"`
+
+	CombinedTarget bool `json:"-"`
+
+	Original interface{} `json:"-"` // Store pointer to provider-specific record object. Used in diffing.
+}
+
+// Providers are responsible for validating or normalizing the data
+// that goes into a RecordConfig.
+
+// Name:
+//    This is the shortname i.e. the NameFQDN without the origin suffix.
+//    It should never have a trailing "."
+//    It should never be null. The apex (naked domain) is stored as "@".
+//    If the origin is "foo.com." and Name is "foo.com", this literally means
+//        the intended FQDN is "foo.com.foo.com." (which may look odd)
+// NameFQDN:
+//    This is the FQDN version of Name.
+//    It should never have a trailiing ".".
+// NOTE: Eventually we will unexport Name/NameFQDN. Please start using
+//     the setters (SetLabel/SetLabelFQDN) and getters (Label/LabelFQDN).
+
+// Target:
+//   This is
+
+// SetLabel sets the .Name/.NameFQDN fields given a short name and origin.
+// origin must not have a trailing dot.
+func (rc *RecordConfig) SetLabel(short, origin string) {
+	if strings.HasSuffix(origin, ".") {
+		panic(fmt.Errorf("origin (%s) is not supposed to end with a dot", origin))
+	}
+
+	short = strings.ToLower(short)
+	origin = strings.ToLower(origin)
+	if short == "" {
+		rc.Name = "@"
+		rc.NameFQDN = origin
+	} else {
+		if short == origin+"." {
+			rc.Name = "@"
+		} else {
+			rc.Name = short
+		}
+		f := dnsutil.AddOrigin(short, origin)
+		rc.NameFQDN = f[0 : len(f)-1]
+	}
+	rc.CheckIntegrity(origin)
+}
+
+// SetLabelFQDN sets the .Name/.NameFQDN fields given a FQDN and origin.
+// fqdn may have a trailing "." but it is not required.
+// origin may not have a trailing dot.
+func (rc *RecordConfig) SetLabelFQDN(fqdn, origin string) {
+
+	if strings.HasSuffix(origin, ".") {
+		panic(fmt.Errorf("origin (%s) is not supposed to end with a dot", origin))
+	}
+	if strings.HasSuffix(fqdn, "..") {
+		panic(fmt.Errorf("fqdn (%s) is not supposed to end with double dots", origin))
+	}
+
+	if strings.HasSuffix(fqdn, ".") {
+		// Trim off a trailing dot.
+		fqdn = fqdn[:len(fqdn)-1]
+	}
+
+	fqdn = strings.ToLower(fqdn)
+	origin = strings.ToLower(origin)
+	rc.Name = dnsutil.TrimDomainName(fqdn, origin)
+	rc.NameFQDN = fqdn
+	rc.CheckIntegrity(origin)
+}
+
+// Label returns the shortname of the label associated with this RecordConfig.
+// It will never end with "."
+// It does not need further shortening (i.e. if it returns "foo.com" and the
+//   domain is "foo.com" then the FQDN is actually "foo.com.foo.com").
+// It will never be "" (the apex is returned as "@").
+func (rc *RecordConfig) Label() string {
```

Comment:
Getters should probably be named `GetLabel` etc... easier to find stuff in autocomplete.